volatility3/framework/plugins/windows/pedump.py

```diff
@@ -0,0 +1,178 @@
+        try:
+            file_handle = self.open(
+                "PE.{:#x}.{:d}.{:#x}.dmp".format(
+                    proc_offset,
+                    pid,
+                    self.config["base"],
+                )
+            )
+
+            dos_header = self.context.object(
+                pe_table_name + constants.BANG + "_IMAGE_DOS_HEADER",
+                offset=self.config["base"],
+                layer_name=layer_name,
+            )
+
+            for offset, data in dos_header.reconstruct():
+                file_handle.seek(offset)
+                file_handle.write(data)
+        except (
+            IOError,
+            exceptions.VolatilityException,
+            OverflowError,
+            ValueError,
+        ) as excp:
+            vollog.debug(f"Unable to dump dll at offset {self.config['base']}: {excp}")
+            return None
+
+        file_handle.close()
+        return file_handle.preferred_filename
+
+    def _dump_kernel(self, _, pe_table_name, session_layers):
+        session_layer_name = modules.Modules.find_session_layer(
+            self.context, session_layers, self.config["base"]
+        )
+
+        if session_layer_name:
+            system_pid = 4
+
+            file_output = self._write_pe(
+                pe_table_name,
+                session_layer_name,
+                0,
+                system_pid,
+            )
+
+            if file_output:
+                yield system_pid, "Kernel", file_output
+        else:
+            vollog.warning(
+                "Unable to find a session layer with the provided base address mapped in the kernel."
+            )
+
+    def _dump_processes(self, kernel, pe_table_name, _):
+        filter_func = pslist.PsList.create_pid_filter(self.config.get("pid", None))
+        kernel = self.context.modules[self.config["kernel"]]
+
+        for proc in pslist.PsList.list_processes(
+            context=self.context,
+            layer_name=kernel.layer_name,
+            symbol_table=kernel.symbol_table_name,
+            filter_func=filter_func,
+        ):
+            pid = proc.UniqueProcessId
+            proc_name = proc.ImageFileName.cast(
+                "string",
+                max_length=proc.ImageFileName.vol.count,
+                errors="replace",
+            )
+            proc_layer_name = proc.add_process_layer()
+
+            file_output = self._write_pe(
+                pe_table_name,
+                proc_layer_name,
+                proc.vol.offset,
+                pid,
+            )
+
+            if file_output:
+                yield pid, proc_name, file_output
+
+    def _generator(self):
+        kernel = self.context.modules[self.config["kernel"]]
+
+        pe_table_name = intermed.IntermediateSymbolTable.create(
+            self.context, self.config_path, "windows", "pe", class_types=pe.class_types
+        )
+
+        if self.config["kernel_module"] and self.config["pid"]:
```

Comment:
This whole flag or pid thing feels... a little ugly?  The main code in this whole plugin is `write_pe` (which should probably be `dump_pe` to be in keeping with other places in the codebase.  This then chooses whether to call one of two methods, but it specifies the method, and then each method takes one or another bogus unneeded parameter.  It feels a very strange way of structuring the code for a choice of two?  I'd also sooner just accept a DLL location (with better error checking in case you're handed a bad pe), and then add `--dump` flags into various other plugins that might want to write a PE?  Either way, if you're specifying the method and the parameters, there's no need to try and make the function signatures match, at that point there's very little being reused...